File: protocols/queen-kafka/compat/rdkafka-ruby-php/php/probe_compression.php
<?php
/**
 * Why every librdkafka client sends UNCOMPRESSED batches to queen-kafka.
 *
 *   php probe-compression.php [bootstrap]
 *
 * compat/README.md documents ONE compression downgrade -- zstd, because librdkafka
 * gates the zstd feature on Fetch v10 and the facade caps Fetch at v6 on purpose.
 * That is real, but it is not the whole story, and this probe exists to show the
 * rest of it: gzip, snappy and lz4 are downgraded too, for a DIFFERENT reason that
 * has nothing to do with Fetch.
 *
 * librdkafka, rdkafka_msgset_writer.c, rd_kafka_msgset_writer_select_MsgVersion():
 *
 *     static const struct { int feature; int16_t ApiVersion; }
 *     compr_req[RD_KAFKA_COMPRESSION_NUM] = {
 *             [RD_KAFKA_COMPRESSION_LZ4]  = {RD_KAFKA_FEATURE_LZ4,  0},
 *             [RD_KAFKA_COMPRESSION_ZSTD] = {RD_KAFKA_FEATURE_ZSTD, 7},
 *     };
 *     if (msetw->msetw_compression &&
 *         (rd_kafka_broker_ApiVersion_supported(
 *              rkb, RD_KAFKAP_Produce, 0,
 *              compr_req[msetw->msetw_compression].ApiVersion, NULL) == -1 || ...))
 *             msetw->msetw_compression = RD_KAFKA_COMPRESSION_NONE;
 *
 * GZIP and SNAPPY are absent from that initialiser, so their entries are zeroed:
 * {feature = 0, ApiVersion = 0}. LZ4's ApiVersion is 0 as well. All three therefore
 * ask the same question -- "does this broker support Produce somewhere in [0, 0]?"
 * -- and rd_kafka_broker_ApiVersion_supported() answers it like this:
 *
 *     else if (ret.MinVer > maxver)  return -1;
 *
 * queen-kafka advertises Produce 3..=9 (versions.rs). MinVer 3 > maxver 0, so the
 * answer is -1 and the codec is dropped. It is the Produce FLOOR that disables
 * gzip/snappy/lz4, not any missing feature -- against a real broker, which
 * advertises Produce from v0, the same call returns 0 and all three compress.
 *
 * zstd is the only one that clears the ApiVersion hurdle (maxver 7 overlaps 3..9)
 * and then fails the SECOND test, the feature bit, which needs Fetch v10.
 *
 * The consequence is invisible in a test that only checks records land: they do
 * land, byte-exact, and librdkafka says so once per DAY per broker at LOG_NOTICE
 * (rd_interval(..., 86400 * 1000 * 1000, 0)). What you lose is the compression.
 */

declare(strict_types=1);

$bootstrap = $argv[1] ?? '127.0.0.1:19092';
$run       = $argv[2] ?? (string) time();

// A payload that any codec would crush: 4 KB of one byte. If a batch really were
// compressed, the produced request would be a fraction of the raw size.
$payload = str_repeat('A', 4096);
$fail    = 0;

printf("bootstrap  %s\n", $bootstrap);
printf("librdkafka %s (php-rdkafka %s)\n\n", (function () {
    ob_start(); phpinfo(INFO_MODULES); $pi = ob_get_clean();
    return preg_match('/librdkafka version \(runtime\)\s*=>\s*(\S+)/', $pi, $m) ? $m[1] : '?';
})(), phpversion('rdkafka'));

foreach (['none', 'gzip', 'snappy', 'lz4', 'zstd'] as $codec) {
    $downgraded = null;
    $conf = new RdKafka\Conf();
    $conf->set('bootstrap.servers', $bootstrap);
    $conf->set('enable.idempotence', 'false');
    $conf->set('acks', 'all');
    $conf->set('compression.codec', $codec);
    $conf->set('debug', 'msg,protocol');
    $conf->set('log_level', '7');
    // The LOG_NOTICE is rate-limited to once per day PER BROKER HANDLE, so each
    // codec gets a brand-new producer or only the first would ever report.
    $conf->setLogCb(function ($k, $l, $f, $m) use (&$downgraded) {
        if (str_contains($m, 'does not support compression type')) {
            $downgraded = trim(substr($m, (int) strpos($m, 'Broker does not support')));
        }
    });
    $delivered = 0;
    $conf->setDrMsgCb(function ($k, RdKafka\Message $m) use (&$delivered) { if (!$m->err) { $delivered++; } });

    $p = new RdKafka\Producer($conf);
    $t = $p->newTopic("codec-probe-$codec-$run");
    for ($i = 0; $i < 32; $i++) { $t->producev(0, RD_KAFKA_MSG_F_BLOCK, $payload, "k$i", []); $p->poll(0); }
    $p->flush(15000);

    $verdict = $codec === 'none'
        ? 'n/a'
        : ($downgraded === null ? 'COMPRESSED' : 'DOWNGRADED to none');
    printf("  %-7s delivered=%-3d %s\n", $codec, $delivered, $verdict);
    if ($downgraded !== null) { printf("          librdkafka: %s\n", $downgraded); }
    if ($delivered !== 32) { $fail++; printf("          FAIL only %d/32 delivered\n", $delivered); }
}

printf("\nEvery codec still DELIVERS -- this costs bandwidth, not correctness.\n");
exit($fail === 0 ? 0 : 1);

File: protocols/queen-kafka/compat/rdkafka-ruby-php/ruby/probe_compression.rb
#!/usr/bin/env ruby
#
# The other side of the librdkafka compression boundary. Run it next to
# php/probe_compression.php and read the two together.
#
#   ruby probe_compression.rb [bootstrap] [runId]
#
# THE FINDING, IN ONE PARAGRAPH
#
# queen-kafka advertises Produce 3..=9 (versions.rs: v0-v2 are the legacy message
# sets nothing sends any more). Real Kafka advertises Produce from v0. librdkafka
# up to and including 2.11.0 decided whether a codec was usable like this
# (rdkafka_msgset_writer.c, rd_kafka_msgset_writer_select_MsgVersion):
#
#     compr_req[LZ4]  = {RD_KAFKA_FEATURE_LZ4,  0};
#     compr_req[ZSTD] = {RD_KAFKA_FEATURE_ZSTD, 7};
#     /* gzip and snappy are absent, so their entries are zeroed: {0, 0} */
#     if (rd_kafka_broker_ApiVersion_supported(rkb, RD_KAFKAP_Produce,
#                                              0, compr_req[codec].ApiVersion,
#                                              NULL) == -1 || ...)
#             msetw->msetw_compression = RD_KAFKA_COMPRESSION_NONE;
#
# For gzip, snappy and lz4 that asks "does the broker support Produce somewhere in
# [0,0]?", and rd_kafka_broker_ApiVersion_supported() answers `if (ret.MinVer >
# maxver) return -1;`. MinVer 3 > 0, so all three codecs are silently dropped --
# not because of any missing feature, purely because of the Produce FLOOR.
#
# librdkafka 2.11.1 replaced that call with rd_kafka_broker_ApiVersion_at_least(),
# which asks the sane question ("is MaxVer >= N?"), and the problem disappears.
#
# So the SAME facade gives two different answers depending on which librdkafka the
# client's packaging happened to link:
#
#   * librdkafka <= 2.11.0  -- gzip, snappy, lz4 AND zstd all downgraded to none.
#   * librdkafka >= 2.11.1  -- gzip, snappy, lz4 compress; only zstd is downgraded,
#                             and that one IS the documented Fetch-v6 cap.
#
# The rdkafka gem ships a PRECOMPILED librdkafka and is currently on the new side.
# php-rdkafka links the SYSTEM librdkafka, and Debian bookworm's is 1.9.2 -- the old
# side. Neither is a correctness problem: every codec still DELIVERS, byte-exact,
# and librdkafka mentions it once per DAY per broker at LOG_NOTICE. What is lost is
# the compression, silently, on exactly the deployments that asked for it.

require "rdkafka"
require "logger"
require "stringio"

BOOTSTRAP = ARGV[0] || "127.0.0.1:19092"
RUN       = ARGV[1] || Time.now.to_i.to_s

# A payload any codec would crush: 4 KB of one byte.
PAYLOAD = ("A" * 4096).freeze

log = StringIO.new
Rdkafka::Config.logger = Logger.new(log)
Rdkafka::Config.logger.level = Logger::DEBUG

puts "bootstrap  #{BOOTSTRAP}"
puts "librdkafka #{Rdkafka::LIBRDKAFKA_VERSION} (rdkafka gem #{Rdkafka::VERSION})"
puts

fail_count = 0
%w[none gzip snappy lz4 zstd].each do |codec|
  mark = log.string.bytesize
  producer = Rdkafka::Config.new(
    "bootstrap.servers"  => BOOTSTRAP,
    "client.id"          => "codec-#{codec}-#{RUN}",
    "enable.idempotence" => "false",
    "acks"               => "all",
    "compression.codec"  => codec,
    "debug"              => "msg",
    "log_level"          => "7"
  ).producer

  handles = (0...32).map do |i|
    producer.produce(topic: "codec-probe-#{codec}-#{RUN}", payload: PAYLOAD, key: "k#{i}", partition: 0)
  end
  producer.flush(15_000)
  # rdkafka 0.29 renamed the seconds-valued `max_wait_timeout:` to
  # `max_wait_timeout_ms:`; try the new name and fall back for older gems.
  delivered = handles.count do |h|
    begin
      h.wait(max_wait_timeout_ms: 20_000)
    rescue ArgumentError
      h.wait(max_wait_timeout: 20)
    rescue StandardError
      nil
    end
  end
  producer.close
  sleep 0.3 # the gem drains its log queue on a background thread

  # The LOG_NOTICE is rate-limited to once per day PER BROKER HANDLE, which is why
  # every codec gets a brand-new producer here -- share one and only the first
  # codec would ever report.
  note = log.string.byteslice(mark..)&.lines&.find { |l| l.include?("does not support compression type") }
  verdict = codec == "none" ? "n/a" : (note ? "DOWNGRADED to none" : "COMPRESSED")
  puts format("  %-7s delivered=%-3d %s", codec, delivered, verdict)
  puts "          librdkafka: #{note.strip.sub(/\A.*?Broker does not/, 'Broker does not')}" if note
  if delivered != 32
    fail_count += 1
    puts "          FAIL only #{delivered}/32 delivered"
  end
end

puts
puts "Every codec still DELIVERS -- this costs bandwidth, not correctness."
exit(fail_count.zero? ? 0 : 1)
